### ISSUE TITLE
Dynamic dispatch in `logpdf`

### ISSUE THREAD

**Issue by cscherrer** [cscherrer/Soss.jl#65]
As of https://github.com/cscherrer/Soss.jl/commit/faa3c24aa14c3de71ff62d22246b530fff68843c,

```julia
julia> @code_warntype logpdf(m(x=x),truth)
Variables
  #self#::Core.Compiler.Const(Distributions.logpdf, false)
  m::Soss.JointDistribution{NamedTuple{(:x,),Tuple{Array{Float64,1}}},NamedTuple{(:x,),T} where T<:Tuple,TypeEncoding(begin
    σ ~ HalfNormal()
    β ~ Normal()
    α ~ Normal()
    yhat = α .+ β .* x
    n = length(x)
    y ~ For(n) do j
            Normal(yhat[j], σ)
        end
end),TypeEncoding(Main)}
  x::NamedTuple{(:x, :σ, :β, :α, :yhat, :n, :y),Tuple{Array{Float64,1},Float64,Float64,Float64,Array{Float64,1},Int64,Array{Float64,1}}}

Body::Any
1 ─ %1 = Soss.from_type($(Expr(:static_parameter, 4)))::Module
│   %2 = Base.getproperty(m, :model)::Model{NamedTuple{(:x,),T} where T<:Tuple,TypeEncoding(begin
    σ ~ HalfNormal()
    β ~ Normal()
    α ~ Normal()
    yhat = α .+ β .* x
    n = length(x)
    y ~ For(n) do j
            Normal(yhat[j], σ)
        end
end),TypeEncoding(Main)}
│   %3 = Base.getproperty(m, :args)::NamedTuple{(:x,),Tuple{Array{Float64,1}}}
│   %4 = Soss._logpdf(%1, %2, %3, x)::Any
└──      return %4
```
Note the `Any` inferences; those are bad ;)

Oddly enough, there's no such issue if you use codegen:

```julia
julia> @code_warntype logpdf(m(x=x),truth, codegen)
Variables
  #self#::Core.Compiler.Const(Distributions.logpdf, false)
  m::Soss.JointDistribution{NamedTuple{(:x,),Tuple{Array{Float64,1}}},NamedTuple{(:x,),T} where T<:Tuple,TypeEncoding(begin
    σ ~ HalfNormal()
    β ~ Normal()
    α ~ Normal()
    yhat = α .+ β .* x
    n = length(x)
    y ~ For(n) do j
            Normal(yhat[j], σ)
        end
end),TypeEncoding(Main)}
  x::NamedTuple{(:x, :σ, :β, :α, :yhat, :n, :y),Tuple{Array{Float64,1},Float64,Float64,Float64,Array{Float64,1},Int64,Array{Float64,1}}}
  #unused#::Core.Compiler.Const(Soss.codegen, false)

Body::Float64
1 ─ %1 = Soss.from_type($(Expr(:static_parameter, 4)))::Module
│   %2 = Base.getproperty(m, :model)::Model{NamedTuple{(:x,),T} where T<:Tuple,TypeEncoding(begin
    σ ~ HalfNormal()
    β ~ Normal()
    α ~ Normal()
    yhat = α .+ β .* x
    n = length(x)
    y ~ For(n) do j
            Normal(yhat[j], σ)
        end
end),TypeEncoding(Main)}
│   %3 = Base.getproperty(m, :args)::NamedTuple{(:x,),Tuple{Array{Float64,1}}}
│   %4 = Soss.codegen(%1, %2, %3, x)::Float64
└──      return %4
```

Both of these use GeneralizedGenerated in very similar ways

**Comment by cscherrer** [comment]
With this model:
```julia
julia> m = @model begin
           x ~ Normal()
       end;
```

The codegen version looks like this
```julia
julia> @code_warntype Soss.codegen(Main, m, NamedTuple(), (x=1.0,))
Variables
  #self#::Core.Compiler.Const(Soss.codegen, false)
  M::Module
  _m::Model{NamedTuple{(),T} where T<:Tuple,TypeEncoding(begin
    x ~ Normal()
end),TypeEncoding(Main)}
  _args::Core.Compiler.Const(NamedTuple(), false)
  _data::NamedTuple{(:x,),Tuple{Float64}}

Body::Float64
1 ─ %1 = (function = (_args, _data;) -> begin
    begin
        begin
            x = (GeneralizedGenerated).getproperty(_data, :x)
            var"##add#483" = 0.0
            var"##add#483" += -0.9189385332046727
            var"##add#483" += begin
                    var"##mul#484" = 1.0
                    var"##mul#484" *= -0.5
                    var"##mul#484" *= begin
                            var"##arg1#486" = x
                            var"##arg2#487" = 2
                            var"##symfunc#485" = (Soss._pow)(var"##arg1#486", var"##arg2#487")
                            var"##symfunc#485"
                        end
                    var"##mul#484"
                end
            var"##add#483"
        end
    end
end)(_args, _data)::Float64
└──      return %1
```

But `_logpdf` is a mess:
```julia
julia> @code_warntype Soss._logpdf(Main, m, NamedTuple(), (x=1.0,))
Variables
  #self#::Core.Compiler.Const(Soss._logpdf, false)
  M::Module
  _m::Model{NamedTuple{(),T} where T<:Tuple,TypeEncoding(begin
    x ~ Normal()
end),TypeEncoding(Main)}
  _args::Core.Compiler.Const(NamedTuple(), false)
  _data::NamedTuple{(:x,),Tuple{Float64}}
  x::Float64
  _ℓ::Any

Body::Any
1 ─      (x = Base.getproperty(_data, :x))
│        (_ℓ = 0.0)
│   %3 = _ℓ::Core.Compiler.Const(0.0, false)::Core.Compiler.Const(0.0, false)
│   %4 = Base.getproperty(M, :logpdf)::Any
│   %5 = Base.getproperty(M, :Normal)::Any
│   %6 = (%5)()::Any
│   %7 = (%4)(%6, x)::Any
│        (_ℓ = %3 + %7)
└──      return _ℓ
```

Which is strange, because from a type inference perspective they don't seem like they'd be so different:
```julia
@gg M function codegen(M::Module, _m::Model, _args, _data)
    f = _codegen(type2model(_m))
    :($f(_args, _data))
end

@gg M function _logpdf(M::Module, _m::Model, _args, _data)  
    type2model(_m) |> sourceLogpdf() |> loadvals(_args, _data)
end
```

**Comment by cscherrer** [comment]
Also extremely odd: When called directly, the latter is actually _faster_!

```julia
julia> @btime $(Soss.codegen)($Main, $m, $(NamedTuple()),$truth)
  1.600 ns (0 allocations: 0 bytes)
-1.019478785641803

julia> @btime $(Soss._logpdf)($Main, $m, $(NamedTuple()),$truth)
compiling
  0.016 ns (0 allocations: 0 bytes)
-1.0194787856418033
```

**Comment by thautwarm** [comment]
should've got fixed by #66
